python/pyarrow/array.pxi

```diff
@@ -1122,14 +1130,8 @@ cdef _array_like_to_pandas(obj, options):
     original_type = obj.type
     name = obj._name
 
-    if obj.type.id == _Type_TIMESTAMP and obj.type.unit != 'ns':
-        # pandas only stores ns data - casting here is faster
-        obj = obj.cast(timestamp('ns'))
-
-    # ARROW-3789(wesm): when converting to DataFrame we coerce
-    # date/timestamp types to nanoseconds. Not consistent but we should
-    # make it so in the future
-    c_options.coerce_temporal_nanoseconds = False
+    # ARROW-3789(wesm): Convert date/timestamp types to datetime64[ns]
+    c_options.coerce_temporal_nanoseconds = True
```

Comment:
I *think* this change is fine: on master, we indeed don't convert eg date64 to datetime64[D] and not to datetime64[ns], in this function (which is used in Array.to_pandas). However, this is then passed to the `Series` constructor, and then pandas will always convert any non-ns datetime64 values to datetime64[ns]. So in practice, for the user, this shouldn't matter if arrow already does this conversion.

This also makes that I can remove the `if` check and `cast()` call just above (which was only done for timestamp, and not for date32/date64), as all datetimelike values are now converted to nanoseconds with this setting (still depending on `date_as_object` for date32/date64 of course)